src/iterator.cr

```diff
@@ -29,6 +29,18 @@ require "./enumerable"
 # (1..10_000_000).each.select(&.even?).map { |x| x * 3 }.first(3).to_a # => [6, 12, 18]
 # ```
 #
+# Due to the lazy computation provided by `Iterator`, emmitted objects may be considered as a
+# potentially infinite stream. Calls that pass to the underlying `Enumerable` module which force
+# eager evaluation of the complete sequence should be used with care. These include `#all?`,
```

Comment:
> should be used with care

I don't think we should let people think that using iterators is dangerous or might lead to bugs. I would simply add a comment like this:

> Because iterators only go forward, when using methods that consume it entirely or partially, like `to_a`, `all?`, `any?` and `size`, subsequent calls to these methods will give a different result because there will be less elements to consume.

The example below is good though 👍

Reply:
Good call. I really like that wording too!